### ISSUE TITLE
Can't set slot access codes

### ISSUE THREAD

**Issue by cmsj** [Yubico/yubikey-manager#12]
Hey :)

I've got a script that I use to nuke/rebuild my yubikey whenever the need arises. I'm trying to port it over to using ykman, but I don't seem to be able to find a way to set the access code for a slot (i.e. the equivalent of adding -oaccess=foo when programming a slot with ykpersonalize.

Is this something that could be added?

Thanks!
Chris

**Comment by dagheyman** [comment]
Hi, thanks for the request.

Yes, that functionality probably could and should be added at some point. I can't give an ETA right now though. Good to know there is a demand for it!

**Comment by aidantwoods** [comment]
So I've managed to lock myself out of my YubiKey, so to speak.

I have OTP mode disabled, while an access code is set. The YubiKey Personalisation Tool (gui and cli) seem to be unable to see the YubiKey with OTP disabled. (I suppose I should bug this, but the tool itself doesn't seem to have been updated in over a year!)

I can't re-enable OTP because changing modes requires no access codes to be present, but I also can't remove the access codes because no tool that can set them can see the YubiKey with OTP off.

After some searching, I found this tool can actually see the YubiKey, but unfortunately not having the ability to set the access code means that this tool too cannot modify my YubiKey.

tl;dr +1 for setting access codes

**Comment by dagheyman** [comment]
>I have OTP mode disabled, while an access code is set.

This shouldn't be possible. When a access code is set, the current mode should be locked, and it shouldn't be possible to disable OTP after that. Can you describe what you did to get in this state? Also, what YubiKey model do you have?

**Comment by aidantwoods** [comment]
So I'll just describe what I was trying to do briefly, so this makes a bit more sense.

Previously I had my YubiKey setup with Yubico OTP on the first slot, and was set to emit my GPG key ID as a static "password" on the second slot. At some point, the OTP mode on the first slot stopped functioning but the U2F remained working and so did the second slot (so acting as a keyboard couldn't have been disabled altogether).

I reprogrammed the first slot the other day, and OTP mode resumed working again, but I actually prefered how it worked before (with OTP disabled on the first slot only – in hindsight, it appears this might've been a bug).

So as for how I got it like this, I was trying to replicate the state it was in before reprogramming the first slot.

> Can you describe what you did to get in this state? 

I used the NEO manager to disable OTP, I then, before unplugging the YubiKey used the Personalisation Tool to set up my YubiKey as before, and configured an access code. After unplugging I haven't been able to remove the access code (since with OTP now disabled the personalisation tool can no longer see the key), neither have I been able to re enable OTP because an access code is configured.

> Also, what YubiKey model do you have?

It's a YubiKey 4 (not a nano)

**Comment by dagheyman** [comment]
@aidantwoods Sorry for the late reply, I've been looking into if there was a way to fix your issue. Unfortunately you seem to actually have locked yourself out from the OTP part of the YubiKey, and I'd suggest you raise a support ticket here: https://www.yubico.com/support/raise-ticket/

Also note that this issue was originally about programming new access codes, but this happens over the OTP transport anyway and wouldn't help in your case, since OTP is disabled and not possible to activate again (due to the access code preventing mode switching).

**Comment by aidantwoods** [comment]
> Unfortunately you seem to actually have locked yourself out from the OTP part of the YubiKey

Ah, well that's not great to hear. I've raised a ticket anyway.

> Also note that this issue was originally about programming new access codes, but this happens over the OTP transport anyway and wouldn't help in your case, since OTP is disabled and not possible to activate again (due to the access code preventing mode switching).

Ah... I wasn't aware so much of the internals that OTP disabled would prevent reprogramming the key outright. Figured it was just a bug in the software that it couldn't see it.

---

Is this something you guys can at least prevent happening with an update to the personalisation tools? Would hate for others to accidentally do this. Luckily I can use (at least slot 1) on my YubiKey in the interim, but could have very easily bricked the thing altogether if I hadn't programmed it before unplugging.

Shall I bug this in the personalisation tool repo?

**Comment by dagheyman** [comment]
@cmsj Actually it seems that programming a slot with an access code is already possible with the current `--access-code` option to the `slot` command. I think the help text for the option needs an update though.

Example:
```
$ ykman slot --access-code 010203040506 static 1 -f
$ ykman slot delete 1 -f
Error: Failed to write to the device. Make sure the device does not have restricted access.
$ ykman slot --access-code 010203040506 delete 1 -f
$ Deleting the configuration of slot 1...
```

**Comment by dagheyman** [comment]
Closing since functionality is in place.
